File: python/03_end_to_end.py
"""
03_end_to_end.py

Submits a build, polls until status === 'complete', then downloads
the HNSW file, loads it, and runs a search.
"""

import os
import sys
import tempfile
import time
import numpy as np
import requests
import hnswlib
from dotenv import load_dotenv

load_dotenv()

api_url     = os.getenv('TT_API_URL')
api_key     = os.getenv('TT_API_KEY')
vectors_url = os.getenv('TT_VECTORS_URL')
dimensions  = int(os.getenv('TT_DIMENSIONS', '0'))
email       = os.getenv('TT_EMAIL')

# Polling interval. Builds typically take minutes-to-hours depending on
# vector count. 60 minutes is a sensible default; for small test sets,
# drop this to 60 (1 minute) or 30.
POLL_INTERVAL_S = 10
MAX_POLLS       = 360

headers = {'X-TensorTensor-API-Key': api_key}


def submit():
    res = requests.post(
        f'{api_url}/api/v1/build',
        json={
            'email': email,
            'vectorsUrl': vectors_url,
            'dimensions': dimensions,
        },
        headers={**headers, 'Content-Type': 'application/json'},
    )
    res.raise_for_status()
    return res.json()['data']


def fetch_status(job_id):
    res = requests.get(f'{api_url}/api/v1/build/{job_id}', headers=headers)
    res.raise_for_status()
    data = res.json().get('data')
    return data[0] if data else None


def download_to_tmp(url):
    fd, tmp_path = tempfile.mkstemp(suffix='.hnsw', prefix=f'tt_{int(time.time())}_')
    os.close(fd)
    with requests.get(url, stream=True) as r:
        r.raise_for_status()
        with open(tmp_path, 'wb') as f:
            for chunk in r.iter_content(chunk_size=1024 * 1024):
                f.write(chunk)
    return tmp_path


def main():
    print('Submitting build...')
    submitted = submit()
    job_id = submitted['jobId']
    print('jobId:', job_id)

    job = None
    for i in range(MAX_POLLS):
        job = fetch_status(job_id)
        print(f'[poll {i + 1}/{MAX_POLLS}] status={job and job.get("status")}')
        if job and job.get('status') in ('complete', 'failed'):
            break
        time.sleep(POLL_INTERVAL_S)

    if not job or job.get('status') != 'complete':
        print('Build did not complete. Final status:', job and job.get('status'), job and job.get('error'), file=sys.stderr)
        sys.exit(1)

    print('Build complete. downloadUrl:', job['downloadUrl'])
    file_path = download_to_tmp(job['downloadUrl'])

    # set to 'l2' for default; for now only 'l2' is supported. future will add support for 'cosine'.
    index = hnswlib.Index(space='l2', dim=dimensions)
    index.load_index(file_path)
    print('Loaded. Element count:', index.get_current_count())

    query = np.random.rand(dimensions).astype('float32')
    k = 5
    labels, distances = index.knn_query(query, k=k)
    print(f'Top {k} nearest neighbors:')
    for i, (lid, dist) in enumerate(zip(labels[0], distances[0]), start=1):
        print(f'  {i}. id={int(lid)}  distance={float(dist):.6f}')


if __name__ == '__main__':
    main()

File: typescript/03-end-to-end.ts
/*
 * 03-end-to-end.ts
 *
 * Submits a build, polls until status === 'complete', then downloads
 * the HNSW file, loads it, and runs a search.
 */

import axios from 'axios';
import * as fs from 'fs';
import * as path from 'path';
import * as os from 'os';
import { HierarchicalNSW } from 'hnswlib-node';
import 'dotenv/config';

interface JobItem {
  jobId: string;
  status: string;
  downloadUrl?: string;
  error?: string;
}

interface JobResponse {
  data: JobItem[];
}

interface SubmittedResponse {
  data: { jobId: string; status: string };
}

const apiUrl     = process.env.TT_API_URL!;
const apiKey     = process.env.TT_API_KEY!;
const vectorsUrl = process.env.TT_VECTORS_URL!;
const dimensions = Number(process.env.TT_DIMENSIONS);
const email      = process.env.TT_EMAIL!;

// Polling interval. Builds typically take minutes-to-hours depending on
// vector count. 60 minutes is a sensible default; for small test sets,
// drop this to 60_000 (1 minute) or 30_000 (30s).
const POLL_INTERVAL_MS = 10 * 1000;
const MAX_POLLS        = 360;

const headers = { 'X-TensorTensor-API-Key': apiKey };

async function submit(): Promise<{ jobId: string; status: string }> {

  const res = await axios.post<SubmittedResponse>(`${apiUrl}/api/v1/build`,
    { email, vectorsUrl, dimensions },
    { headers: { ...headers, 'Content-Type': 'application/json' } },
  );
  return res.data.data;

}

async function fetchStatus(jobId: string): Promise<JobItem | undefined> {

  const res = await axios.get<JobResponse>(`${apiUrl}/api/v1/build/${jobId}`, { headers });
  return res.data.data?.[0];

}

function sleep(ms: number): Promise<void> {

  return new Promise(r => setTimeout(r, ms));

}

async function downloadToTmp(url: string): Promise<string> {

  const file = path.join(os.tmpdir(), `tt_${Date.now()}.hnsw`);
  const writer = fs.createWriteStream(file);
  const res = await axios.get(url, { responseType: 'stream' });
  await new Promise<void>((resolve, reject) => {
    res.data.pipe(writer);
    writer.on('finish', () => resolve());
    writer.on('error', reject);
  });
  return file;

}

async function main(): Promise<void> {

  console.log('Submitting build...');
  const submitted = await submit();
  const jobId = submitted.jobId;
  console.log('jobId:', jobId);

  let job: JobItem | undefined;
  for (let i = 0; i < MAX_POLLS; i++){
    job = await fetchStatus(jobId);
    console.log(`[poll ${i + 1}/${MAX_POLLS}] status=${job?.status}`);
    if (job?.status === 'complete' || job?.status === 'failed') break;
    await sleep(POLL_INTERVAL_MS);
  }

  if (job?.status !== 'complete'){
    console.error('Build did not complete. Final status:', job?.status, job?.error);
    process.exit(1);
  }

  console.log('Build complete. downloadUrl:', job.downloadUrl);
  const filePath = await downloadToTmp(job.downloadUrl!);

  // set to 'l2' for default; for now only 'l2' is supported. future will add support for 'cosine'.
  const index = new HierarchicalNSW('l2', dimensions);
  index.readIndexSync(filePath);
  console.log('Loaded. Element count:', index.getCurrentCount());

  const query: number[] = Array.from({ length: dimensions }, () => Math.random());
  const k = 5;
  const result = index.searchKnn(query, k);
  console.log(`Top ${k} nearest neighbors:`);
  result.neighbors.forEach((id: number, i: number) => {
    console.log(`  ${i + 1}. id=${id}  distance=${result.distances[i].toFixed(6)}`);
  });

}

main().catch((err: any) => {
  console.error('error:', err.response?.data || err.message);
  process.exit(1);
});
